lib/matplotlib/backends/registry.py

```diff
@@ -0,0 +1,56 @@
+from enum import Enum
+
+
+class BackendFilter(Enum):
+    INTERACTIVE = 0
+    NON_INTERACTIVE = 1
+
+
+class BackendRegistry:
+    """
+    Registry of backends available within Matplotlib.
+
+    This is the single source of truth for available backends.
+    """
+    # Built-in backends are those which are included in the Matplotlib repo.
+    # A backend with name 'name' is located in the module
+    # f'matplotlib.backends.backend_{name.lower()}'
+
+    # The capitalized forms are needed for ipython at present; this may
+    # change for later versions.
+    _BUILTIN_INTERACTIVE = [
+        "GTK3Agg", "GTK3Cairo", "GTK4Agg", "GTK4Cairo",
+        "MacOSX",
+        "nbAgg",
+        "QtAgg", "QtCairo", "Qt5Agg", "Qt5Cairo",
+        "TkAgg", "TkCairo",
+        "WebAgg",
+        "WX", "WXAgg", "WXCairo",
+    ]
+    _BUILTIN_NOT_INTERACTIVE = [
+        "agg", "cairo", "pdf", "pgf", "ps", "svg", "template",
+    ]
+    _GUI_FRAMEWORK_TO_BACKEND_MAPPING = {
+        "qt": "qtagg",
+        "gtk3": "gtk3agg",
+        "gtk4": "gtk4agg",
+        "wx": "wxagg",
+        "tk": "tkagg",
+        "macosx": "macosx",
+        "headless": "agg",
+    }
+
+    def backend_for_gui_framework(self, framework):
+        return self._GUI_FRAMEWORK_TO_BACKEND_MAPPING.get(framework)
+
+    def list_builtin(self, filter_=None):
+        if filter_ == BackendFilter.INTERACTIVE:
+            return self._BUILTIN_INTERACTIVE
+        elif filter_ == BackendFilter.NON_INTERACTIVE:
+            return self._BUILTIN_NOT_INTERACTIVE
+
+        return self._BUILTIN_INTERACTIVE + self._BUILTIN_NOT_INTERACTIVE
+
+
+# Singleton
+backend_registry = BackendRegistry()
```

Comment:
I lean towards the explicit and verbose approach, but maybe there is some middle ground. If I keep the files named as they are but import into `matplotlib.backends.__init__` using `from .registry import BackendFilter, backend_registry` then the standard usage becomes
```python
from matplotlib.backends import backend_registry
```
rather than
```python
from matplotlib.backends.registry import backend_registry
```
removing some verbosity.

This has the added advantage that you cannot import the `BackendRegistry` class in the same way which is good as nobody should need the class, only the singleton instance.

I have implemented this in commit 8f10a01c.